ndcube/meta.py

```diff
@@ -0,0 +1,373 @@
+    def comments(self):
+        return self._comments
+
+    @property
+    def axes(self):
+        return self._axes
+
+    @property
+    def shape(self):
+        return self._data_shape
+
+    def add(self, name, value, comment=None, axis=None, overwrite=False):
+        """Add a new piece of metadata to instance.
+
+        Parameters
+        ----------
+        name: `str`
+            The name/label of the metadata.
+
+        value: Any
+            The value of the metadata. If axes input is not None, this must have the
+            same length/shape as those axes as defined by ``self.shape``.
+
+        comment: `str` or `None`
+            Any comment associated with this metadata. Set to None if no comment desired.
+
+        axis: `int`, iterable of `int`, or `None`
+            The axis/axes with which the metadata is linked. If not associated with any
+            axis, set this to None.
+
+        overwrite: `bool`, optional
+            If True, overwrites the entry of the name name if already present.
+        """
+        if name in self.keys() and overwrite is not True:
+            raise KeyError(f"'{name}' already exists. "
+                           "To update an existing metadata entry set overwrite=True.")
+        if comment is not None:
+            self._comments[name] = comment
+        if axis is not None:
+            axis = self._sanitize_axis_value(axis, value, name)
+            self._axes[name] = axis
+        elif name in self._axes:
+            del self._axes[name]
+        # This must be done after updating self._axes otherwise it may error.
+        self.__setitem__(name, value)
+
+    def remove(self, name):
+        if name in self._comments:
+            del self._comments[name]
+        if name in self._axes:
+            del self._axes[name]
+        del self[name]
+
+    def __setitem__(self, key, val):
+        axis = self.axes.get(key, None)
+        if axis is not None:
+            if _not_scalar(val):
+                axis_shape = tuple(self.shape[axis])
+                if not _is_grid_aligned(val, axis_shape) and not _is_axis_aligned(val, axis_shape):
+                    raise TypeError(
+                        f"{key} is already associated with axis/axes {axis}. val must therefore "
+                        f"must either have same length as number associated axes ({len(axis)}), "
+                        f"or the same shape as associated data axes {tuple(self.shape[axis])}. "
+                        f"val shape = {val.shape if hasattr(val, 'shape') else (len(val),)}\n"
+                        "We recommend using the 'add' method to set values.")
+        super().__setitem__(key, val)
+
+    def __getitem__(self, item):
+        # There are two ways to slice:
+        # by key, or
+        # by typical python numeric slicing API,
+        # i.e. slice the each piece of metadata associated with an axes.
+
+        if isinstance(item, str):
+            return super().__getitem__(item)
```

Comment:
I would suggest we split out the  cube-like slicing from `__getitem__` into a property so you do something like `mymeta.slice[10:20]` if you want to do cube-like slicing.